hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieWriteClient.java

```diff
@@ -1193,6 +1194,17 @@ protected final HoodieTable initTable(WriteOperationType operationType, Option<S
       default:
     }
 
+    // Keep hoodie.properties updated from write config
+    HoodieTableConfig tableConfig = metaClient.getTableConfig();
+    if (!StringUtils.isNullOrEmpty(config.getTableName()) && !config.getTableName().equals(tableConfig.getTableName())) {
+      LOG.info("Table name for dataset has changed from " + tableConfig.getTableName() + " to " + config.getTableName());
+      tableConfig.setTableName(config.getTableName());
+      HoodieTableConfig.update(metaClient.getFs(), new Path(metaClient.getMetaPath()), tableConfig.getProps());
```

Comment:
Usually table config should have high priority because it is shared by all the writers, not sure whether we should allow the change of table name by just a change of write config.

A more general solution for table name changing is from the SQL `ALTER TABLE` statement.

Reply:
In that case, there is not much use of table name in the HoodieWriteConfig. 
1. If any code path depends on table name from HoodieWriteConfig then that would be using the wrong table name (different from hoodie.properties)
2. If no code path is using table name from HoodieWriteConfig then we may remove it
3. The writers seems to be using the table name from HoodieWriteConfig and readers from hoodie.properties. There is a disconnect here.

Reply:
Yeah, I was just addressing some converns from the hign level, like the user habits comparing to normal RDBMS, usually people rename the table first though catalog or CLI, then restart the write pipeline.